### ISSUE TITLE
Remove PinnableBufferCache from corefx

### ISSUE THREAD

**Issue by stephentoub** [dotnet/corefx#15761]
The only remaining use of PinnableBufferCache in corefx is in SslStream, which uses two different instantiations of it to get read and write buffers.  I suggest we replace this usage with ArrayPool, so as to consolidate pools, and if doing so results in any regression, incorporate whatever we need from PinnableBufferCache into ArrayPool to address it (e.g. right now ArrayPool doesn't explicitly pay attention to generations, though it's not clear whether that negatively impacts it for these scenarios).  Then we can delete the internal PinnableBufferCache copy from corefx.

cc: @geoffkizer, @jkotas, @brianrob

**Comment by brianrob** [comment]
Sounds reasonable to me.  Fragmentation is really the thing to watch when doing this.  However, know that it will take a fairly large workload to see the effects, both in terms of fragmentation as well as the cost in GC time of handling a very large number of objects that need to be moved during compaction.

You won't get pushback from me on wanting to merge pooling solutions, but just be aware that there is pretty significant scale required to see these effects.

cc: @vancem

**Comment by vancem** [comment]
I am OK with this approach.

**Comment by benaadams** [comment]
**tl;dr** The main issue is handling depletion+overflow scenarios; otherwise all buffers will be Gen2 anyway. In general, shouldn't be any issue. However, new api to override default pool where usage pattern is known? 

----
The only issue I can see is from a fast pool sharing with a slow pool. Fast pool I'd define as getting a buffer, using it in a cpu operation then returning. Slow pool I'd define as getting a buffer, sitting on it for a while, then returning it. (Obviously the pooling operations would both want to be "fast").

Some read operations would be slow; so a network read where its in the keep-alive state so the buffer is just sat on for a while. 4.5.x Websockets (and maybe current client) where it they would hold on to the buffer for the lifetime of the socket would again be slow.

The slow users would empty the pool of buffers. The fast users would then allocate new ones; when the slow users came to return theirs, they would overflow and be discarded.

However saying that... all the buffers would probably quickly reach Gen2 buffers so there might not be any difference between them; and no reason to differentiate this way.

What may be useful and since the shared pool is getting much more use, is some way to override the shared pool in certain situations; where there is application knowledge. 

So if the pool has 50 arrays per bucket; but I know I'll have 500 on the go at any one time; or 500 stream copies happen on average; then I might want to force that particular type of thing off the shared pool and onto is own pool and have it sized differently. Additionally that different pool may want to be resused between 2 or more scenarios. 

Not sure if its a per op, or a global operation type setting e.g. `FileStream.Pool.set` vs `CopyAsync(...., pool)` though the first is probably an easier place to start?

**Comment by benaadams** [comment]
Hmm... equally it might just as easily be solved by allowing the shared pool defaults to be increased (via startup configuration would be fine)

**Comment by benaadams** [comment]
Raised issue for that https://github.com/dotnet/corefx/issues/15773

**Comment by karelz** [comment]
This is not blocking 2.0. It's nice to have, but does not impact customers directly.

**Comment by stephentoub** [comment]
Fixed by #23572
